File: internal/service/dashboard_svc.go
package service

import (
	"context"
	"fmt"
	"time"

	"github.com/zajca/zfaktury/internal/domain"
	"github.com/zajca/zfaktury/internal/repository"
)

// DashboardData holds all aggregated data for the dashboard view.
type DashboardData struct {
	RevenueCurrentMonth  domain.Amount
	ExpensesCurrentMonth domain.Amount
	UnpaidCount          int
	UnpaidTotal          domain.Amount
	OverdueCount         int
	OverdueTotal         domain.Amount
	MonthlyRevenue       []repository.MonthlyAmount
	MonthlyExpenses      []repository.MonthlyAmount
	RecentInvoices       []repository.RecentInvoice
	RecentExpenses       []repository.RecentExpense
}

// DashboardService provides business logic for the dashboard.
type DashboardService struct {
	repo repository.DashboardRepo
}

// NewDashboardService creates a new DashboardService.
func NewDashboardService(repo repository.DashboardRepo) *DashboardService {
	return &DashboardService{repo: repo}
}

// GetDashboard fetches all dashboard data for the current month and year.
func (s *DashboardService) GetDashboard(ctx context.Context) (*DashboardData, error) {
	now := time.Now()
	year := now.Year()
	month := int(now.Month())

	revenue, err := s.repo.RevenueCurrentMonth(ctx, year, month)
	if err != nil {
		return nil, fmt.Errorf("fetching revenue: %w", err)
	}

	expenses, err := s.repo.ExpensesCurrentMonth(ctx, year, month)
	if err != nil {
		return nil, fmt.Errorf("fetching expenses: %w", err)
	}

	unpaidCount, unpaidTotal, err := s.repo.UnpaidInvoices(ctx)
	if err != nil {
		return nil, fmt.Errorf("fetching unpaid invoices: %w", err)
	}

	overdueCount, overdueTotal, err := s.repo.OverdueInvoices(ctx)
	if err != nil {
		return nil, fmt.Errorf("fetching overdue invoices: %w", err)
	}

	monthlyRevenue, err := s.repo.MonthlyRevenue(ctx, year)
	if err != nil {
		return nil, fmt.Errorf("fetching monthly revenue: %w", err)
	}

	monthlyExpenses, err := s.repo.MonthlyExpenses(ctx, year)
	if err != nil {
		return nil, fmt.Errorf("fetching monthly expenses: %w", err)
	}

	recentInvoices, err := s.repo.RecentInvoices(ctx, 5)
	if err != nil {
		return nil, fmt.Errorf("fetching recent invoices: %w", err)
	}

	recentExpenses, err := s.repo.RecentExpenses(ctx, 5)
	if err != nil {
		return nil, fmt.Errorf("fetching recent expenses: %w", err)
	}

	return &DashboardData{
		RevenueCurrentMonth:  revenue,
		ExpensesCurrentMonth: expenses,
		UnpaidCount:          unpaidCount,
		UnpaidTotal:          unpaidTotal,
		OverdueCount:         overdueCount,
		OverdueTotal:         overdueTotal,
		MonthlyRevenue:       monthlyRevenue,
		MonthlyExpenses:      monthlyExpenses,
		RecentInvoices:       recentInvoices,
		RecentExpenses:       recentExpenses,
	}, nil
}

File: rust/crates/zfaktury-core/src/service/dashboard_svc.rs
use std::sync::Arc;

use chrono::Datelike;
use zfaktury_domain::{Amount, DomainError};

use crate::repository::traits::DashboardRepo;
use crate::repository::types::{MonthlyAmount, RecentExpense, RecentInvoice};

/// Aggregated dashboard data.
pub struct DashboardData {
    pub revenue_current_month: Amount,
    pub expenses_current_month: Amount,
    pub unpaid_count: i64,
    pub unpaid_total: Amount,
    pub overdue_count: i64,
    pub overdue_total: Amount,
    pub monthly_revenue: Vec<MonthlyAmount>,
    pub monthly_expenses: Vec<MonthlyAmount>,
    pub recent_invoices: Vec<RecentInvoice>,
    pub recent_expenses: Vec<RecentExpense>,
}

/// Service for the dashboard view.
pub struct DashboardService {
    repo: Arc<dyn DashboardRepo + Send + Sync>,
}

impl DashboardService {
    pub fn new(repo: Arc<dyn DashboardRepo + Send + Sync>) -> Self {
        Self { repo }
    }

    pub fn get_dashboard(&self) -> Result<DashboardData, DomainError> {
        let now = chrono::Local::now();
        let year = now.date_naive().year();
        let month = now.date_naive().month() as i32;

        let revenue = self.repo.revenue_current_month(year, month)?;
        let expenses = self.repo.expenses_current_month(year, month)?;
        let (unpaid_count, unpaid_total) = self.repo.unpaid_invoices()?;
        let (overdue_count, overdue_total) = self.repo.overdue_invoices()?;
        let monthly_revenue = self.repo.monthly_revenue(year)?;
        let monthly_expenses = self.repo.monthly_expenses(year)?;
        let recent_invoices = self.repo.recent_invoices(5)?;
        let recent_expenses = self.repo.recent_expenses(5)?;

        Ok(DashboardData {
            revenue_current_month: revenue,
            expenses_current_month: expenses,
            unpaid_count,
            unpaid_total,
            overdue_count,
            overdue_total,
            monthly_revenue,
            monthly_expenses,
            recent_invoices,
            recent_expenses,
        })
    }
}
